*** common/conf.h	Wed Sep 30 14:34:38 1992
--- common/conf.h	Wed Nov 18 18:43:44 1992
***************
*** 4,9 ****
--- 4,22 ----
   */
  
  /*
+  * Extensions to NNTP RFC977 (I.Lea 18-11-92)
+  */
+ 
+ #define XOVERVIEW 		/* retieve .overview index files from server */
+ #define XINDEX			/* retieve tin index files from server       */
+ #define XMOTD			/* retieve message of the day from server    */
+ #define XUSER			/* log clients username to nntp logfile      */
+ 
+ #define XINDEX_DIR		"/usr/spool/news/.index"
+ #define XMOTD_FILE		"/usr/lib/news/motd"
+ #define SUBSCRIPTIONS_FILE	"/usr/lib/news/subscriptions"
+ 
+ /*
   * Compile time options.
   */
  
*** common/conf.h.dist		Wed Sep 30 14:34:35 1992
--- common/conf.h.dist		Wed Nov 18 18:44:47 1992
***************
*** 4,13 ****
   */
  
  /*
   * Compile time options.
   */
    
- 
  #undef	ALONE		/* True if we're running without inetd */
  
  #ifdef ALONE
--- 4,25 ----
   */
  
  /*
+  * Extensions to NNTP RFC977 (I.Lea 18-11-92)
+  */
+ 
+ #define XOVERVIEW 		/* retieve .overview index files from server */
+ #define XINDEX			/* retieve tin index files from server       */
+ #define XMOTD			/* retieve message of the day from server    */
+ #define XUSER			/* log clients username to nntp logfile      */
+ 
+ #define XINDEX_DIR		"/usr/spool/news/.index"
+ #define XMOTD_FILE		"/usr/lib/news/motd"
+ #define SUBSCRIPTIONS_FILE	"/usr/lib/news/subscriptions"
+ 
+ /*
   * Compile time options.
   */
    
  #undef	ALONE		/* True if we're running without inetd */
  
  #ifdef ALONE
*** common/nntp.h	Wed Sep 30 14:34:38 1992
--- common/nntp.h	Wed Nov 18 19:19:00 1992
***************
*** 38,43 ****
--- 38,52 ----
  #define	OK_GOODBYE	205	/* Closing connection */
  #define	OK_GROUP	211	/* Group selected */
  #define	OK_GROUPS	215	/* Newsgroups follow */
+ #ifdef XMOTD
+ #define	OK_XMOTD	217	/* Motd (message of the day) file follows */
+ #endif
+ #ifdef XINDEX
+ #define	OK_XINDEX	218	/* Tin style group index file follows */
+ #endif
+ #ifdef XOVERVIEW
+ #define	OK_XOVERVIEW	219	/* .overview style group index file follows */
+ #endif
  #define	OK_ARTICLE	220	/* Article (head & body) follows */
  #define	OK_HEAD		221	/* Head follows */
  #define	OK_BODY		222	/* Body follows */
***************
*** 57,62 ****
--- 66,80 ----
  #define	ERR_GOODBYE	400	/* Have to hang up for some reason */
  #define	ERR_NOGROUP	411	/* No such newsgroup */
  #define	ERR_NCING	412	/* Not currently in newsgroup */
+ #ifdef XMOTD
+ #define	ERR_XMOTD	417	/* No motd (message of the day) file */
+ #endif
+ #ifdef XINDEX
+ #define	ERR_XINDEX	418	/* No tin style index file for newsgroup */
+ #endif
+ #ifdef XOVERVIEW
+ #define	ERR_XOVERVIEW	419	/* No .overview style index file for newsgroup */
+ #endif
  #define	ERR_NOCRNT	420	/* No current article selected */
  #define	ERR_NONEXT	421	/* No next article in this group */
  #define	ERR_NOPREV	422	/* No previous article in this group */
