*** ar.c.old	Sun Oct  4 00:10:50 1992
--- ar.c	Sun Oct  4 00:11:43 1992
***************
*** 1209,1215 ****
  make_tempname (name)
       char *name;
  {
! #ifdef USG
    /* sigh. 14 character filenames are *wonderful*, just *wonderful*. */
    char *p = basename (name);
    char *q, *r;
--- 1209,1215 ----
  make_tempname (name)
       char *name;
  {
! #if defined(USG) || defined(SHORT_FILENAME)
    /* sigh. 14 character filenames are *wonderful*, just *wonderful*. */
    char *p = basename (name);
    char *q, *r;
